internal/store/store.go

```diff
@@ -484,35 +489,110 @@ func (s *Store) PushCompose(composeID uuid.UUID, bp *blueprint.Blueprint, checks
-
-	pipeline, err := s.distro.Pipeline(bp, repos, checksums, arch, composeType, size)
+	pipelineStruct, err := s.distro.Pipeline(bp, repos, checksums, arch, composeType, size)
 	if err != nil {
 		return err
 	}
-	s.change(func() error {
-		s.Composes[composeID] = Compose{
-			QueueStatus: "WAITING",
-			Blueprint:   bp,
-			Pipeline:    pipeline,
-			OutputType:  composeType,
-			Targets:     targets,
-			JobCreated:  time.Now(),
-			Size:        size,
+	// FIXME: handle or comment this possible error
+	_ = s.change(func() error {
+		s.Composes[composeID] = compose.Compose{
+			Blueprint: bp,
+			ImageBuilds: []compose.ImageBuild{
+				{
+					QueueStatus: common.IBWaiting,
+					Pipeline:    pipelineStruct,
+					ImageType:   imageType,
+					Targets:     targets,
+					JobCreated:  time.Now(),
+					Size:        size,
+				},
+			},
 		}
 		return nil
 	})
 	s.pendingJobs <- Job{
-		ComposeID:  composeID,
-		Distro:     s.distro.Name(),
-		Pipeline:   pipeline,
-		Targets:    targets,
-		OutputType: composeType,
+		ComposeID: composeID,
+		Distro:    s.distro.Name(),
+		Pipeline:  pipelineStruct,
+		Targets:   targets,
+		ImageType: composeType,
 	}
 
 	return nil
 }
 
+// PushComposeRequest is an alternative to PushCompose which does not assume a pre-defined distro, as such it is better
+// suited for RCM API and possible future API that would respect the fact that we can build any distro and any arch
+func (s *Store) PushComposeRequest(request common.ComposeRequest) error {
+	// This should never happen and once distro.Pipeline is refactored this check will go away
+	arch, exists := request.Arch.ToString()
+	if !exists {
+		panic("fatal error, arch should exist but it does not")
+	}
+	distroString, exists := request.Distro.ToString()
+	if !exists {
+		panic("fatal error, distro should exist but it does not")
+	}
+
+	distroStruct := s.distroRegistry.GetDistro(distroString)
+	if distroStruct == nil {
+		panic("fatal error, distro should exist but it is not in the registry")
+	}
+
+	// This will be a list of imageBuilds that will be submitted to the state channel
+	var imageBuilds []compose.ImageBuild
+
+	// TODO: remove this
+	if len(request.RequestedImages) > 1 {
+		panic("Multiple image requests are not yet properly implemented")
+	}
+
+	for _, imageRequest := range request.RequestedImages {
+		// TODO: handle custom upload targets
+		// TODO: this requires changes in the Compose Request struct
+		// TODO: implment support for no checksums
+		// This should never happen and once distro.Pipeline is refactored this check will go away
+		imgTypeCompatStr, exists := imageRequest.ImgType.ToCompatString()
+		if !exists {
+			panic("fatal error, image type should exist but it does not")
+		}
+		pipelineStruct, err := distroStruct.Pipeline(&request.Blueprint, request.Repositories, nil, arch, imgTypeCompatStr, 0)
+		if err != nil {
+			return err
+		}
+
+		s.pendingJobs <- Job{
+			ComposeID: request.ComposeID,
+			Distro:    s.distro.Name(),
```

Comment:
Why not use the `request.Distro` value here? I think we should get rid of `s.distro` altogether because we will want to make cross-distro builds in the future.

Reply:
Thanks, this is a clear mistake.

Reply:
Fixed